File: 895-maximum-frequency-stack/895-maximum-frequency-stack.py
class FreqStack:

    def __init__(self):
        self.stackHmap = {}
        self.freqHmap = {}
        self.maxFreq = -1

    def push(self, val: int) -> None:
        if val in self.freqHmap:
            self.freqHmap[val] += 1
            pos = self.freqHmap[val]
        else:
            self.freqHmap[val] = 1
            pos = self.freqHmap[val]
        if pos in self.stackHmap:
            self.stackHmap[pos].append(val)
        else:
            self.stackHmap[pos] = [val]
        
        self.maxFreq = max(self.maxFreq, pos)

    def pop(self) -> int:
        ans = self.stackHmap[self.maxFreq].pop()
        if len(self.stackHmap[self.maxFreq]) == 0:
            self.maxFreq -= 1
        self.freqHmap[ans] -= 1
        return ans


# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()

File: 895-maximum-frequency-stack/895-maximum-frequency-stack.java
class FreqStack {
        Map<Integer, Integer> freqMap;
        Map<Integer, Stack<Integer>> stackMap;
        Integer maxFreq;
    
    public FreqStack() {
        freqMap = new HashMap();
        stackMap = new HashMap();
        maxFreq = -1;
    }
    
    public void push(int val) {
        int f = freqMap.getOrDefault(val, 0) + 1;
        freqMap.put(val, f);
        maxFreq = Math.max(maxFreq, f);
        
        if(!stackMap.containsKey(f)) {
            stackMap.put(f, new Stack());
        }
        stackMap.get(f).push(val);
    }
    
    public int pop() {
        int ans = stackMap.get(maxFreq).pop();
        freqMap.put(ans, freqMap.get(ans)-1);
        
        if(stackMap.get(maxFreq).isEmpty()) {
            maxFreq--;
        }
        return ans;
    
    }
}

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack obj = new FreqStack();
 * obj.push(val);
 * int param_2 = obj.pop();
 */